tests/test_site_apportionment/test_site_apportionment.py

```diff
@@ -634,3 +636,127 @@ def test_sort_rows_missing_cols(self, caplog):
+                [49900000407, 2.0, 202212, 'test', 'AA', 'D', np.nan, np.nan, 100.0, 0.0],
+                [49900000576, 1.0, 202212, 'TMI', 'AA', 'C', np.nan, np.nan, np.nan, 0.0],
+                [49900000960, 0.0, 202212, 'R', 'AA', np.nan, 1.0, np.nan, np.nan, np.nan],
+                [49900000960, 1.0, 202212, 'R', 'AA', 'C', 23.0, np.nan, np.nan, np.nan],
+                [49900001029, 1.0, 202212, 'TMI', 'I', 'C', 5.0, np.nan, np.nan, 833.3333]]
+        
+        columns = ['reference', 'instance', 'period', 'imp_marker', '201', 
+                   '200', 'pg_numeric', '601', '602', '210']
+        
+        return pd.DataFrame(data=data, columns=columns)
+
+    @pytest.fixture(autouse=True)
+    def set_attrs(self):
+        """Set class attributes to pass to functions for testing."""
+        self.imp_markers_to_keep = ["R", "TMI", "CF", "MoR", "constructed"]
+        self.groupby_cols = ["reference", "period"]
+        self.code_cols = ["201", "200", "pg_numeric"]
+        # excluding postcode_harmonised as it is irrelevant to this test
+        self.site_cols = ["instance", "601", "602"]
+        self.value_cols = ["210"]
+
+    def test_create_category_df_raises(self, category_df_input):
+        """Defensive tests for create_category_df."""
+        # simple test passing wrong input type to a param
+        with pytest.raises(TypeError, match="can only concatenate list.*int.*list"):
+            create_category_df(category_df_input,
+                                self.imp_markers_to_keep,
+                                category_df_input.columns,
+                                self.groupby_cols,
+                                6,
+                                self.site_cols,
+                                self.value_cols)
+
+    def test_create_category_df_on_pass(self, category_df_input):
+        """General tests for create_category_df."""
+        output = create_category_df(category_df_input,
+                                    self.imp_markers_to_keep,
+                                    category_df_input.columns,
+                                    self.groupby_cols,
+                                    self.code_cols,
+                                    self.site_cols,
+                                    self.value_cols)
+    
+        # assert correct number of rows
+        assert len(output) == 5, "Unexpected number of records in output."
+        # assert group code nan dropped
+        GC_NAN_CONDS = (output["201"].isin(["I"])
+                        &
+                        output["reference"].isin([49900000408]))
+        assert len(output[GC_NAN_CONDS]) == 0, "nan group code not removed"
+        # assert that bad imp marker removed
+        assert len((output[output["reference"].isin([49900000407])])) == 0,(
+            "Bad imp marker not removed"
+        ) 
+        # assert value_col (210) as e expected
+        expected = [243600.0379, 0.0, 0.0, 0.0, 833.3333]
```

Comment:
I'd still like to see an expected output dataframe please, with the columns listed. Thanks.